akka-stream/src/main/scala/akka/stream/impl/PhasedFusingActorMaterializer.scala

```diff
@@ -504,7 +525,23 @@ private final case class SavedIslandData(islandGlobalOffset: Int, lastVisitedOff
       islandTracking.allNestedIslandsReady()
 
       if (Debug) println("--- Finished materialization")
-      matValueStack.peekLast().asInstanceOf[Mat]
+      matValueStack.peekLast() match {
+        case FlattenMatHolder(f: Future[Mat @unchecked], t) ⇒
+          f.value match {
+            case Some(Success(m))  ⇒ m
+            case Some(Failure(ex)) ⇒ throw new Exception("Flattened materialized value failed!", ex)
+            case None ⇒
+              // last resort, await
+              val start = System.currentTimeMillis()
+              val mat = Await.result(f, t)
```

Comment:
Isn't the problem, as @jrudolph pointed out, that no matter how short the expected blocking is that we risk deadlock due to starvation?

If you have max 10 threads in the pool (not using FJP which will create new threads for Await), and materialize 100 such streams at the same time. Are you guaranteed that all will complete or is there any risk that all 10 threads will be stuck in the Await?

Reply:
While I do think we need a facility like this, since the need of surfacing a materialized value out of a Logic strictly, without the need for Futures is not that uncommon I think, esp for surfacing the ActorRefs etc, backing out of this for now. The deadlock scenario I can't really see in reality, materialization failures could happen though, blocking is bounded (could be in small timing) and materialization itself already is blocking too.

Anyway, `back to the Future[]` API and will re-explore getting rid of it again separately.
Will move the API out of Sink / Source, as I really don't want the `Future` one to be here and later have to break it.